File: CSES/3. Dynamic Programming/CSES-1635_Coin Combinations I.cpp
#include <bits/stdc++.h>
using namespace std;
const int MOD = 1e9 + 7;
#define endl '\n'

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    int n, x;
    cin >> n >> x;
    vector<int> coins(n);
    for (int i = 0; i < n; i++) cin >> coins[i];
    sort(coins.begin(), coins.end());
    vector<int> dp(x + 1, 0);
    dp[0] = 1;
    // 刷表法
    // for (int i = 0; i <= x; i++) {
    //     for (int c : coins) {
    //         if (i + c > x) break;
    //         dp[i + c] += dp[i];
    //         dp[i + c] %= MOD;
    //     }
    // }
    // 填表法
    for (int i = 0; i <= x; i++) {
        for (int c : coins) {
            if (i - c < 0) break;
            dp[i] += dp[i - c];
            dp[i] %= MOD;
        }
    }
    cout << dp[x] << endl;
    return 0;
}

File: CSES/3. Dynamic Programming/CSES-1635_Coin Combinations I.py
"""
    DP
    Python 用填表法會 TLE，要用刷表法；C++ 則是兩種都可以
"""
n, x = map(int, input().split())
coins = list(map(int, input().split()))
coins.sort() # 這裡排序是為了剪枝
MOD = int(1e9 + 7)

dp = [0] * (x + 1)
dp[0] = 1
# 填表會 TLE
# for i in range(x + 1):
#     for c in coins:
#         if i - c < 0:
#             break
#         dp[i] += dp[i - c]
#         dp[i] %= MOD
for i in range(x + 1):
    for c in coins:
        if i + c > x: # 剪枝
            break
        dp[i + c] += dp[i]
        dp[i + c] %= MOD
print(dp[x])